components/layout_2020/flow/root.rs

```diff
@@ -229,15 +200,165 @@ impl FragmentTreeRoot {
+        let mut iterator = FragmentContentBoxIterator::new(requested_node);
+        self.iterate_through_fragments(&mut iterator);
+        let rect = &iterator.rect;
+
+        Rect::new(
+            Point2D::new(
+                Au::from_f32_px(rect.origin.x.px()),
+                Au::from_f32_px(rect.origin.y.px()),
+            ),
+            Size2D::new(
+                Au::from_f32_px(rect.size.width.px()),
+                Au::from_f32_px(rect.size.height.px()),
+            ),
+        )
+    }
+
+    pub fn get_border_dimensions_for_node(&self, requested_node: OpaqueNode) -> Rect<i32> {
+        let mut iterator = FragmentBorderDimensionsIterator::new(requested_node);
+        self.iterate_through_fragments(&mut iterator);
+        iterator.border_dimensions
+    }
+}
+
+/// A top-down fragment border box iteration handler.
+trait FragmentIterator {
+    /// The operation to perform.
+    fn process(&mut self, fragment: &Fragment, containing_block: &PhysicalRect<Length>) -> bool;
+}
+
+struct FragmentContentBoxIterator {
+    node_address: OpaqueNode,
+    rect: PhysicalRect<Length>,
+}
+
+impl FragmentContentBoxIterator {
+    fn new(node_address: OpaqueNode) -> FragmentContentBoxIterator {
+        FragmentContentBoxIterator {
+            node_address,
+            rect: PhysicalRect::zero(),
+        }
+    }
+}
+
+impl FragmentIterator for FragmentContentBoxIterator {
+    fn process(&mut self, fragment: &Fragment, containing_block: &PhysicalRect<Length>) -> bool {
+        let fragment_relative_rect = match fragment {
+            Fragment::Box(fragment) if fragment.tag == self.node_address => fragment
+                .border_rect()
+                .to_physical(fragment.style.writing_mode, &containing_block),
+            Fragment::Text(fragment) if fragment.tag == self.node_address => fragment
+                .rect
+                .to_physical(fragment.parent_style.writing_mode, &containing_block),
+            Fragment::Box(_) | Fragment::Text(_) | Fragment::Image(_) | Fragment::Anonymous(_) => {
+                return true
+            },
+        };
+
+        self.rect = fragment_relative_rect.translate(containing_block.origin.to_vector());
+        return false;
+    }
+}
+
+struct FragmentBorderDimensionsIterator {
+    node_address: OpaqueNode,
+    border_dimensions: Rect<i32>,
+}
+
+impl FragmentBorderDimensionsIterator {
+    fn new(node_address: OpaqueNode) -> FragmentBorderDimensionsIterator {
+        FragmentBorderDimensionsIterator {
+            node_address,
+            border_dimensions: Rect::zero(),
+        }
+    }
+}
+
+impl FragmentIterator for FragmentBorderDimensionsIterator {
+    fn process(&mut self, fragment: &Fragment, containing_block: &PhysicalRect<Length>) -> bool {
+        let (style, border_box) = match fragment {
+            Fragment::Box(fragment) if fragment.tag == self.node_address => {
+                (&fragment.style, fragment.border_rect())
```

Comment:
Wouldn’t using `padding_rect` here make the computation below easier?

Reply:
Yep. It's much simpler using padding_rect here.